### PR TITLE
e2e: refactor metrics test to use NSD and WI

### PR DESCRIPTION
**Pull request by shoenig**
This PR overhauls the metrics e2e suite which has been failing for like a year. It swaps out the use of Consul in favor of Nomad native service discovery and workload identity.

Basically it runs a handful of random jobs which produce metrics, then uses Prometheus to gather those metrics. Caddy is used to expose the Prometheus API to be reachable from the test runner. The little [nomad-holepunch](https://github.com/shoenig/nomad-holepunch) thing is used as a side car to enable Prometheus to access the Nomad API using workload identity. It's also used as a service job to represent each Nomad client in the Nomad service registry.

### CODE REVIEW

**Comment by shoenig** [comment]
Spot check against e2e

```
nomad/e2e/metrics on e2e-metrics-metrics-metrics
➜ go test -v
=== RUN   TestMetrics
    metrics_test.go:52: tweaking podman registry auth files ...
    metrics_test.go:56: running metrics job cpustress ...
    metrics_test.go:60: running metrics job nomadagent ...
    metrics_test.go:64: running metrics job prometheus ...
    metrics_test.go:68: running metrics job pythonhttp ...
    metrics_test.go:72: running metrics job caddy ...
    metrics_test.go:76: let the metrics collect for a bit (10s) ...
    metrics_test.go:79: measuring alloc metrics ...
    metrics_test.go:153: expose prometheus http address http://3.86.25.98:9999
    metrics_test.go:174: query for metric nomad_client_allocs_memory_usage{exported_job="nomadagent-928"}
    metrics_test.go:185: -> empty vector, will try again in 5 seconds
    metrics_test.go:174: query for metric nomad_client_allocs_memory_usage{exported_job="nomadagent-928"}
    metrics_test.go:185: -> empty vector, will try again in 5 seconds
    metrics_test.go:174: query for metric nomad_client_allocs_memory_usage{exported_job="nomadagent-928"}
    metrics_test.go:185: -> empty vector, will try again in 5 seconds
    metrics_test.go:174: query for metric nomad_client_allocs_memory_usage{exported_job="nomadagent-928"}
    metrics_test.go:185: -> empty vector, will try again in 5 seconds
    metrics_test.go:174: query for metric nomad_client_allocs_memory_usage{exported_job="nomadagent-928"}
    metrics_test.go:185: -> empty vector, will try again in 5 seconds
    metrics_test.go:174: query for metric nomad_client_allocs_memory_usage{exported_job="nomadagent-928"}
    metrics_test.go:174: query for metric nomad_client_allocs_cpu_user{exported_job="cpustress-348"}
    metrics_test.go:174: query for metric nomad_client_allocs_cpu_allocated{exported_job="pythonhttp-071"}
    metrics_test.go:94: measuring client metrics ...
    metrics_test.go:153: expose prometheus http address http://3.86.25.98:9999
    metrics_test.go:174: query for metric nomad_client_allocated_memory{node_id="0169fb1a-8631-a65d-dc26-45c7f47dbf69"}
    metrics_test.go:174: query for metric sum(nomad_client_host_cpu_user{node_id="0169fb1a-8631-a65d-dc26-45c7f47dbf69"})
    metrics_test.go:174: query for metric nomad_client_host_memory_used{node_id="0169fb1a-8631-a65d-dc26-45c7f47dbf69"}
    metrics_test.go:174: query for metric nomad_client_uptime{node_id="0169fb1a-8631-a65d-dc26-45c7f47dbf69"}
    metrics_test.go:174: query for metric nomad_client_allocated_memory{node_id="c60934dc-ca04-dd43-76cb-478b59a7fe56"}
    metrics_test.go:174: query for metric sum(nomad_client_host_cpu_user{node_id="c60934dc-ca04-dd43-76cb-478b59a7fe56"})
    metrics_test.go:174: query for metric nomad_client_host_memory_used{node_id="c60934dc-ca04-dd43-76cb-478b59a7fe56"}
    metrics_test.go:174: query for metric nomad_client_uptime{node_id="c60934dc-ca04-dd43-76cb-478b59a7fe56"}
    metrics_test.go:174: query for metric nomad_client_allocated_memory{node_id="5123e03c-b00d-523a-1c7c-e1f09404c0e5"}
    metrics_test.go:174: query for metric sum(nomad_client_host_cpu_user{node_id="5123e03c-b00d-523a-1c7c-e1f09404c0e5"})
    metrics_test.go:174: query for metric nomad_client_host_memory_used{node_id="5123e03c-b00d-523a-1c7c-e1f09404c0e5"}
    metrics_test.go:174: query for metric nomad_client_uptime{node_id="5123e03c-b00d-523a-1c7c-e1f09404c0e5"}
    metrics_test.go:174: query for metric nomad_client_allocated_memory{node_id="5733b7f2-f742-dc07-18af-73132c844e56"}
    metrics_test.go:174: query for metric sum(nomad_client_host_cpu_user{node_id="5733b7f2-f742-dc07-18af-73132c844e56"})
    metrics_test.go:174: query for metric nomad_client_host_memory_used{node_id="5733b7f2-f742-dc07-18af-73132c844e56"}
    metrics_test.go:174: query for metric nomad_client_uptime{node_id="5733b7f2-f742-dc07-18af-73132c844e56"}
--- PASS: TestMetrics (77.29s)
PASS
ok      github.com/hashicorp/nomad/e2e/metrics  77.299s
```

File: e2e/metrics/input/caddy.hcl [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+{
+  auto_https off
+  http_port 3000
+}
+http:// {
+{{ $allocID := env "NOMAD_ALLOC_ID" -}}
+{{ range nomadService 1 $allocID "prometheus" }}
+  reverse_proxy {{ .Address }}:{{ .Port }}
+{{ end }}
+}
```
Comment (lgfa29) [inline e2e/metrics/input/caddy.hcl:69]:
```suggestion
http:// {
  reverse_proxy {{ range nomadService 1 $allocID "prometheus" }}{{ .Address }}:{{ .Port }} {{end}}
}
```

Not that it matters much in this case, but we could let Caddy handle the load balancing. I'm also not too familar with Caddyfiles, so I hope this is correct 😅

File: e2e/metrics/input/cpustress.hcl [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      name     = "stress"
+      tags     = ["cpu"]
+    }
+
+    task "cpustress" {
+      driver = "pledge"
+
+      config {
+        command  = "stress"
+        args     = ["--cpu", "1", ]
```
Comment (lgfa29) [inline e2e/metrics/input/cpustress.hcl:35]:
```suggestion
        args     = ["--cpu", "1"]
```

File: e2e/metrics/input/cpustress.hcl [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    service {
+      provider = "nomad"
+      name     = "stress"
+      tags     = ["cpu"]
+    }
+
+    task "cpustress" {
+      driver = "pledge"
```
Comment (lgfa29) [inline e2e/metrics/input/cpustress.hcl:31]: Nice!

File: e2e/metrics/input/prometheus.hcl [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      check {
+        type     = "http"
+        path     = "/-/healthy"
+        interval = "5s"
+        timeout  = "2s"
+      }
+    }
+
+    # run a private holepunch instance in this group network
+    # so prometheus can access the nomad api for service disco
```
Comment (lgfa29) [inline e2e/metrics/input/prometheus.hcl:47]: Nice x2

**Review by lgfa29 — APPROVED** [review]
Great work! It seems like a nice mix of task drivers, would make sense to use Docker for one of the Podman task to add even more flavours?

File: e2e/metrics/input/caddy.hcl (hunk 1 above)
Comment (shoenig) [reply e2e/metrics/input/caddy.hcl:69]: I gave this a try and it failed ... probably not going to spend the time to investigate, so many other test failures to look into :sweat:

**Comment by shoenig** [comment]
> would make sense to use Docker for one of the Podman task to add even more flavours?

Yeah for sure, we can look into expanding this especially once we get the windows client back

**Pull request merged**